File: services/tracking/src/shared/cache/keys.py
"""Cache key construction.

## Why every response key carries TWO identities

Every response key carries BOTH identities — `cognito_sub` and `user_id`.
`cognito_sub` is the ownership key every user-scoped read filters by
(`services/tracking/CLAUDE.md` §5b); `user_id` is the internal `usr_` id. Both
travel so a key is unambiguous under either identity model, and so the per-user
index below can be reconstructed from either.

## Why a builder may answer `None`

`user_id` is resolved lazily, over gRPC to Users, and that resolution is allowed
to fail: `log_identity._resolve_quietly` swallows `UnknownUserError`,
`grpc.RpcError` and everything else, because enriching a log line must never fail
a request. So a fully authenticated caller can reach a handler with
`user_id is None`.

Formatting that `None` into a key would produce the literal string
`tracking:order:v1:<sub>:None:<order_id>` — which is a DIFFERENT key per sub, so
it does not leak across users by itself. But the `None` segment is a lie about
what the entry is scoped by, and the per-user index keyed on the same `None`
would collapse. Answering `None` instead makes the route skip caching for that
request entirely: it pays a MISS, serves from MySQL, and writes nothing. A
request that cannot be keyed correctly is not cached at all.

## Why the list key is a hash

`order_ids` is an arbitrary caller-supplied list of up to `MAX_BATCH_ORDER_IDS`
(100) ids. Keying on the raw list would make the key length proportional to the
request and the key SPACE combinatorial. Sorting and deduplicating first, then
hashing, collapses every ordering and every repetition of one set onto one
fixed-length key — which is both a cardinality bound and a hit-rate improvement,
since two clients asking for the same orders in different orders now share an
entry.
"""

from __future__ import annotations

import hashlib

#: Bumped when a cached DTO's shape changes, which mass-invalidates every entry
#: of that shape without touching Redis: the old keys simply stop being read and
#: expire on their own TTL.
VERSION = "v1"

#: The number of colon-separated segments that make up a key's PREFIX — the only
#: part of a key that may appear in a span attribute, a metric dimension or a log
#: line. Everything after it is identity.
_PREFIX_SEGMENTS = 3


class CacheKeys:
    """Namespace for the key builders. No state; never instantiated."""

    @staticmethod
    def tracking_order(
        cognito_sub: str, user_id: str | None, order_id: str
    ) -> str | None:
        """Key for `GET /v1/trackings/{order_id}`, or None if unkeyable."""
        if not user_id:
            return None
        return f"tracking:order:{VERSION}:{cognito_sub}:{user_id}:{order_id}"

    @staticmethod
    def tracking_list(
        cognito_sub: str, user_id: str | None, order_ids: list[str]
    ) -> str | None:
        """Key for `GET /v1/trackings?order_ids=`, or None if unkeyable.

        Normalizes (sort + dedup) BEFORE hashing, so `?order_ids=b,a` and
        `?order_ids=a,b,a` are one key.
        """
        if not user_id:
            return None
        digest = _hash_order_ids(order_ids)
        return f"tracking:list:{VERSION}:{cognito_sub}:{user_id}:{digest}"

    @staticmethod
    def identity(cognito_sub: str) -> str:
        """Key for the `cognito_sub -> user_id` mapping.

        Never `None`: this is the cache consulted to OBTAIN a `user_id`, so it
        cannot require one. Keyed on the sub alone.
        """
        return f"identity:sub-to-user:{VERSION}:{cognito_sub}"

    @staticmethod
    def user_index(cognito_sub: str, user_id: str) -> str:
        """Key of the Redis SET holding this user's live response keys.

        Required because a list key embeds a HASH of an arbitrary id list and
        therefore cannot be reconstructed at invalidation time. `KEYS` and `SCAN`
        are the wrong answer: both are O(N) over the whole keyspace, and `KEYS`
        blocks the server while it runs.
        """
        return f"tracking:index:{VERSION}:{cognito_sub}:{user_id}"

    @staticmethod
    def prefix_of(key: str) -> str:
        """The telemetry-safe prefix: everything up to and including `v1`.

        A full key carries `cognito_sub` and `user_id`. A span is an export
        destination like any other, and a CloudWatch dimension VALUE is
        cardinality the account is billed for, so neither ever sees more than
        this.
        """
        return ":".join(key.split(":")[:_PREFIX_SEGMENTS])


def _hash_order_ids(order_ids: list[str]) -> str:
    """Normalize then hash: sorted, deduplicated, newline-joined, sha256.

    `sorted(set(...))` is the normalization; the newline join is a separator that
    cannot appear inside an order id, so `["ab", "c"]` and `["a", "bc"]` cannot
    collide. Truncated to 16 hex characters — 64 bits, which for a keyspace of at
    most a few million live list entries makes a collision negligible, while
    keeping the key short enough to read in `redis-cli`.

    sha256 rather than `hash()`: Python's built-in string hash is salted per
    process (PYTHONHASHSEED), so two replicas would compute DIFFERENT keys for
    the same request and the cache would never hit across them.
    """
    normalized = "\n".join(sorted(set(order_ids)))
    return hashlib.sha256(normalized.encode("utf-8")).hexdigest()[:16]

File: services/tracking-go/internal/adapter/redis/keys.go
// Package redis is the cache adapter: keys, gateway, identity cache, and eviction.
// CONTRACT: Do NOT treat cognitoSub as canonical; it is raw x-user-id and may be
// either a Cognito sub or usr_ id. Assuming only the sub leaves deleted-account
// entries readable until TTL expiry. Response keys carry both that raw identity
// and the resolved usr_ id; unresolved user_id makes a request unkeyable rather
// than collapsing callers into an empty segment. List IDs are normalized and
// hashed to keep equivalent sets on one bounded key.
// See [[tracking-service-design]]
package redis

import (
	"crypto/sha256"
	"encoding/hex"
	"sort"
	"strings"
)

// Version is bumped when a cached DTO's shape changes, which mass-invalidates
// every entry of that shape without touching Redis: the old keys simply stop
// being read and expire on their own TTL.
const Version = "v1"

// prefixSegments is how many colon-separated segments make up a key's PREFIX —
// the only part of a key that may appear in a span attribute, a metric dimension
// or a log line. Everything after it is identity.
const prefixSegments = 3

// digestLength in hex characters. 64 bits, which for a keyspace of at most a few
// million live list entries makes a collision negligible, while keeping the key
// short enough to read in redis-cli.
const digestLength = 16

// TrackingOrderKey builds the key for GET /v1/trackings/{order_id}. The second
// return is false when the request is unkeyable — see the package docstring.
func TrackingOrderKey(cognitoSub, userID, orderID string) (string, bool) {
	if userID == "" {
		return "", false
	}
	return "tracking:order:" + Version + ":" + cognitoSub + ":" + userID + ":" + orderID, true
}

// TrackingListKey builds the key for GET /v1/trackings?order_ids=. Normalizes
// (sort + dedup) BEFORE hashing, so ?order_ids=b,a and ?order_ids=a,b,a are one
// key.
func TrackingListKey(cognitoSub, userID string, orderIDs []string) (string, bool) {
	if userID == "" {
		return "", false
	}
	return "tracking:list:" + Version + ":" + cognitoSub + ":" + userID + ":" + hashOrderIDs(orderIDs), true
}

// IdentityKey builds the key for the identifier -> user_id mapping.
//
// Never answers "no key": this is the cache consulted to OBTAIN a user_id, so it
// cannot require one. That identity is whatever the caller sent, not necessarily
// a Cognito sub — an invalidator must not assume this key is reachable under the
// canonical sub only.
func IdentityKey(cognitoSub string) string {
	return "identity:sub-to-user:" + Version + ":" + cognitoSub
}

// UserIndexKey builds the key of the Redis SET holding this user's live response
// keys.
// CONTRACT: Do NOT replace the per-user index with KEYS or SCAN; whole-keyspace
// work blocks or scales every invalidation with cache size. The first segment is
// raw x-user-id, so one person can own multiple indexes.
// See [[tracking-service-design]]
func UserIndexKey(cognitoSub, userID string) string {
	return "tracking:index:" + Version + ":" + cognitoSub + ":" + userID
}

// PrefixOf returns the telemetry-safe prefix: everything up to and including v1.
//
// A full key carries cognito_sub and user_id. A span is an export destination
// like any other, and a CloudWatch dimension VALUE is cardinality the account is
// billed for, so neither ever sees more than this.
func PrefixOf(key string) string {
	parts := strings.Split(key, ":")
	if len(parts) <= prefixSegments {
		return key
	}
	return strings.Join(parts[:prefixSegments], ":")
}

// hashOrderIDs normalizes then hashes: sorted, deduplicated, newline-joined,
// sha256, truncated.
//
// CONTRACT: Do NOT use a process-seeded runtime hash; replicas would compute
// different keys and never share cache hits. Newlines prevent ambiguous joins,
// and sha256 keeps the digest deterministic.
// See [[tracking-service-design]]
func hashOrderIDs(orderIDs []string) string {
	seen := make(map[string]struct{}, len(orderIDs))
	unique := make([]string, 0, len(orderIDs))
	for _, id := range orderIDs {
		if _, dup := seen[id]; dup {
			continue
		}
		seen[id] = struct{}{}
		unique = append(unique, id)
	}
	sort.Strings(unique)

	sum := sha256.Sum256([]byte(strings.Join(unique, "\n")))
	return hex.EncodeToString(sum[:])[:digestLength]
}
